src/Microsoft.Health.Fhir.Core.UnitTests/Performance/ResourceHandlerPerformanceTests.cs

```diff
@@ -0,0 +1,101 @@
+        private Observation resource;
+
+        [GlobalSetup]
+        public void Initialize()
+        {
+            _dataStore = new InMemoryDataStore();
+            _conformanceProvider = Substitute.For<ConformanceProviderBase>();
+
+            _rawResourceFactory = Substitute.For<RawResourceFactory>(new FhirJsonSerializer());
+            _resourceWrapperFactory = Substitute.For<IResourceWrapperFactory>();
+            _resourceWrapperFactory
+                .Create(Arg.Any<Resource>(), Arg.Any<bool>())
+                .Returns(x => CreateResourceWrapper(x.ArgAt<Resource>(0), x.ArgAt<bool>(1)));
+
+            _conformanceStatement = CapabilityStatementMock.GetMockedCapabilityStatement();
+            CapabilityStatementMock.SetupMockResource(_conformanceStatement, ResourceType.Observation, null);
+            var observationResource = _conformanceStatement.Rest.First().Resource.Find(x => x.Type == ResourceType.Observation);
+            observationResource.ReadHistory = false;
+            observationResource.UpdateCreate = true;
+            observationResource.Versioning = CapabilityStatement.ResourceVersionPolicy.Versioned;
+
+            CapabilityStatementMock.SetupMockResource(_conformanceStatement, ResourceType.Patient, null);
+            var patientResource = _conformanceStatement.Rest.First().Resource.Find(x => x.Type == ResourceType.Patient);
+            patientResource.ReadHistory = true;
+            patientResource.UpdateCreate = true;
+            patientResource.Versioning = CapabilityStatement.ResourceVersionPolicy.VersionedUpdate;
+
+            _conformanceProvider.GetCapabilityStatementAsync().Returns(_conformanceStatement);
+            var lazyConformanceProvider = new Lazy<IConformanceProvider>(() => _conformanceProvider);
+
+            var collection = new ServiceCollection();
+
+            collection.Add(x => new CreateResourceHandler(_dataStore, lazyConformanceProvider, _resourceWrapperFactory)).Singleton().AsSelf().AsImplementedInterfaces();
+            collection.Add(x => new UpsertResourceHandler(_dataStore, lazyConformanceProvider, _resourceWrapperFactory)).Singleton().AsSelf().AsImplementedInterfaces();
+            collection.Add(x => new GetResourceHandler(_dataStore, lazyConformanceProvider, _resourceWrapperFactory)).Singleton().AsSelf().AsImplementedInterfaces();
+            collection.Add(x => new DeleteResourceHandler(_dataStore, lazyConformanceProvider, _resourceWrapperFactory)).Singleton().AsSelf().AsImplementedInterfaces();
+
+            ServiceProvider provider = collection.BuildServiceProvider();
+            _mediator = new Mediator(type => provider.GetService(type));
+
+            resource = Samples.GetDefaultObservation();
+            resource.Id = null;
+        }
+
+        [Benchmark]
+        public async Task GivenAFhirMediator_WhenSavingAResourceWithNoId_ThenAnIdShouldBeAssigned()
+        {
+            await _mediator.UpsertResourceAsync(resource);
```

Comment:
It's not immediately obvious to me what this is measuring. Maybe add an explanation?